### PR TITLE
[PT2] FBC

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

- Impanated FBC for experimental tracing
- Save graph to GraphModelWrapper, to using graph like for NNCFNetwork
- Add ConstantLayerAttributes to constant node
- Check is_experimental_torch_tracing_enabled inside patch_torch_operators, to support [optimum-intel ](https://github.com/huggingface/optimum-intel/blob/f601b8b1fda4477ed9f9e4293cf3c9d5cec4ad1b/optimum/intel/openvino/__init__.py#L49)

### Related tickets

152996

### CODE REVIEW

File: nncf/experimental/torch2/function_hook/nncf_graph/nncf_graph_builder.py [hunk 1]
```diff
@@ -228,17 +230,17 @@ class GraphModelWrapper:
     """
     A class that wraps a PyTorch model with examples inputs and provides an interface
     to build a computational graph of the model.
-
-    :param model: The PyTorch model to be wrapped.
-    :param example_input: A tuple of example input for the model.
     """
 
-    def __init__(self, model: nn.Module, example_input: Any) -> None:
+    def __init__(self, model: nn.Module, example_input: Any, graph: Optional[PTNNCFGraph] = None) -> None:
```
Comment (alexsu52) [inline nncf/experimental/torch2/function_hook/nncf_graph/nncf_graph_builder.py:235]: I couldn't find the place in the code where `GraphModelWrapper` is created with graph. I would say that it looks like as not use functionality and this generally contradicts the fact that you have a `reset` method.

**Review by alexsu52 — CHANGES_REQUESTED** [review]

File: nncf/torch/model_graph_manager.py [hunk 2]
```diff
@@ -51,9 +52,7 @@ def find_const_node_in_constant_subgraph(node: NNCFNode, graph: NNCFGraph) -> Op
     :return: The constant node found within the subgraph, or None if no constant node is found.
     """
     if node.metatype == om.PTNoopMetatype or node.node_type in om.QUANTIZE_NODE_TYPES:
-        prev_nodes = graph.get_previous_nodes(node)
-        if len(prev_nodes) != 1:
-            return None
+        prev_nodes = [e.from_node for e in graph.get_input_edges(node)]
         return find_const_node_in_constant_subgraph(prev_nodes[0], graph)
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/torch/model_graph_manager.py:56]: For experimental tracing FQ nodes have edges to constant node

File: nncf/torch/model_graph_manager.py [hunk 3]
```diff
@@ -118,7 +117,7 @@ def get_module_by_name(module_name: str, model: torch.nn.Module) -> torch.nn.Mod
     return curr_module
 
 
-def get_const_data(const_node: NNCFNode, model: NNCFNetwork) -> torch.Tensor:
+def get_const_data(const_node: NNCFNode, model: nn.Module) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/torch/model_graph_manager.py:120]: It expects nn.Module, NNCFNetwork interfaces is not used.

File: nncf/experimental/torch2/function_hook/nncf_graph/nncf_graph_builder.py (hunk 1 above)
Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/function_hook/nncf_graph/nncf_graph_builder.py:235]: yeap, need to remove it

File: nncf/torch/model_graph_manager.py [hunk 4]
```diff
@@ -189,42 +188,40 @@ def is_node_with_fused_bias(node: NNCFNode, nncf_graph: NNCFGraph) -> bool:
     return bias is not None
 
 
-def get_fused_bias_value(node: NNCFNode, model: NNCFNetwork) -> Optional[torch.Tensor]:
+def get_fused_bias_value(node: NNCFNode, nncf_graph: NNCFGraph, model: NNCFNetwork) -> Optional[torch.Tensor]:
     """
     Returns the bias tensor for the node or for potential fused node.
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/torch/model_graph_manager.py:193]: updated

File: nncf/experimental/torch2/function_hook/extractor.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+def apply_args_to_kwargs(
+    args: Sequence[Any], kwargs: Dict[str, Any], indexed_args: List[Tuple[int, str]]
+) -> Dict[str, Any]:
+    """
+    Applies the given arguments and keyword arguments to a dictionary of keyword arguments.
+
+    :param args: The positional arguments.
+    :param kwargs: The keyword arguments.
+    :param indexed_args: The list of pairs of indexes and names.
+
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch2/function_hook/extractor.py:64]:
```suggestion
```

File: nncf/experimental/torch2/function_hook/extractor.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        elif arg_name in kwargs:
+            args_dict[arg_name] = kwargs[arg_name]
+
+    return args_dict
+
+
+def extract_bn(model: nn.Module, graph: PTNNCFGraph, node: NNCFNode) -> ExtractedFunc:
+    """
+    Extract batch_norm operation.
+    If source modules inhered from nn.BatchNorm1d, nn.BatchNorm2d, or nn.BatchNorm3d return torch BatchNorm module.
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch2/function_hook/extractor.py:80]:
```suggestion
    If source modules inhered from nn.BatchNorm1d, nn.BatchNorm2d, or nn.BatchNorm3d, return torch BatchNorm module.
```

File: nncf/experimental/torch2/model_transformer.py [hunk 7]
```diff
@@ -34,8 +38,9 @@ class PT2ModelTransformer(ModelTransformer[GraphModelWrapper]):
     def __init__(self, model: GraphModelWrapper):
         super().__init__(model)
 
-        self._command_transformation_ordered_pairs = [
-            (PT2InsertionCommand, self._apply_insertion_transformation),
+        self._command_transformation_ordered_pairs: TRANSFORMATION_PAIRS = [
+            (PT2InsertionCommand, self._apply_insertion_transformations),
+            (PTBiasCorrectionCommand, self._apply_bias_correction_transformations),
         ]
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch2/model_transformer.py:44]: Just for a bit of saferty
```suggestion
        self._command_transformation_ordered_pairs: TRANSFORMATION_PAIRS = (
            (PT2InsertionCommand, self._apply_insertion_transformations),
            (PTBiasCorrectionCommand, self._apply_bias_correction_transformations),
        )
```

File: tests/torch2/function_hook/quantization/test_fast_bias_correction.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        nncf_graph = model.get_graph()
+        for node in nncf_graph.get_all_nodes():
+            if not is_node_with_fused_bias(node, nncf_graph):
+                continue
+            bias_value = get_fused_bias_value(node, nncf_graph, model.model).cpu()
+            # TODO(AlexanderDokuchaev): return atol=0.0001 after fix 109189
+            assert torch.all(torch.isclose(bias_value, ref_bias, atol=0.02)), f"{bias_value} != {ref_bias}"
+            return
+        msg = "Not found node with bias"
+        raise ValueError(msg)
```
Comment (daniil-lyakhov) [inline tests/torch2/function_hook/quantization/test_fast_bias_correction.py:92]: Can we reuse this method from `TestTorchFBCAlgorithm`?

Comment (AlexanderDokuchaev) [reply tests/torch2/function_hook/quantization/test_fast_bias_correction.py]: yes, done same for tests/torch/ptq/test_fast_bias_correction.py

File: nncf/experimental/torch2/model_transformer.py (hunk 7 above)
Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/model_transformer.py]: done

File: nncf/experimental/torch2/function_hook/extractor.py (hunk 6 above)
Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/function_hook/extractor.py]: Good catch 👍 but problem is not in comma.
This line is copypast that should be removed, now batch norm module is not required

**Comment by AlexanderDokuchaev** [comment]
Model | Metric value | Compr. time | Bias correction time | RAM MiB
-- | -- | -- | -- | --
torchvision/resnet18 | 0.69152 | 0:00:19 | 0:00:00 | 1384
timm/wide_resnet50_2 | 0.81164 | 0:01:01 | 0:00:01 | 2703
timm/vgg11 | 0.68754 | 0:00:53 | 0:00:00 | 4072
timm/tf_inception_v3 | 0.77624 | 0:01:27 | 0:00:01 | 2061
timm/resnest14d | 0.74876 | 0:00:28 | 0:00:00 | 1447
timm/regnetx_002 | 0.68476 | 0:00:35 | 0:00:00 | 1314
timm/mobilenetv3_small_050 | 0.42922 (+0.00252) | 0:00:37 | 0:00:00 | 1252
timm/mobilenetv2_050 | 0.6528 | 0:00:42 | 0:00:00 | 1286
timm/inception_resnet_v2 | 0.803 | 0:03:57 | 0:00:04 | 3399
timm/hrnet_w18 | 0.7725 | 0:04:46 | 0:00:01 | 2867
timm/efficientnet_lite0 | 0.75186 | 0:00:42 | 0:00:01 | 1393
timm/efficientnet_b0 | 0.77158 (+0.00116) | 0:01:02 | 0:00:00 | 1511
timm/dpn68 | 0.77158 | 0:01:09 | 0:00:00 | 1679

1) mobilenetv3_small_050 and efficientnet_b0:
found bug with extractor of batch_norm on develop version - fix will be in another PR (need to disable gradients)
2) dla34: endless working quantization solver, need to align tracing of concat layer  - fix will be in another PR

**Review by daniil-lyakhov — APPROVED** [review]

**Review by alexsu52 — COMMENTED** [review]
LGTM

File: nncf/torch/graph/operator_metatypes.py [hunk 9]
```diff
@@ -727,8 +728,15 @@ class PTBatchNormMetatype(PTOperatorMetatype):
         NamespaceTarget.ATEN: ["_native_batch_norm_legit_no_training", "cudnn_batch_norm"],
     }
     subtypes = [PTModuleBatchNormMetatype]
-    weight_port_ids = [3]
-    bias_port_id = 4
+
+    if is_experimental_torch_tracing_enabled():
```
[the remark this answers is not in the document]
Comment (anzr299) [reply nncf/torch/graph/operator_metatypes.py:732]: Yes, I can handle this case in my PR

**Review by alexsu52 — APPROVED** [review]
